### PR TITLE
oauth-ruby overhaul

### PR DESCRIPTION
**Pull request by kevinhughes27**
The original version of this gem used hand-rolled HMAC, which was discovered during a code review to add this gem to our PCI environment. While we still wanted the convenience of using a gem to sign our OAuth 1.0a requests, the current digest algorithm in use was unacceptable. See: http://www.ruby-doc.org/stdlib-2.0/libdoc/digest/rdoc/Digest/HMAC.html (which was backported originally by hand.)

See commit history for improvements to code structure as well as use of OpenSSL implementations for all digest and hashing algorithms, and some pruning to implementations not specified in the OAuth 1.0a specification. 

paired with @bizla 

for review
@fw42 @mutemule
/cc @Shopify/platform @Shopify/payments

### CODE REVIEW

File: lib/oauth/signature/hmac/sha2.rb [hunk 1]
```diff
@@ -1,8 +0,0 @@
-require 'oauth/signature/hmac/base'
-
-module OAuth::Signature::HMAC
```
Comment (fw42) [inline lib/oauth/signature/hmac/sha2.rb]: why delete this and not do the same as with sha1.rb?

**Comment by fw42** [comment]
I think this looks much better. We should keep SHA2 and MD5 and submit this upstream, imho.

**Comment by kevinhughes27** [comment]
@fw42 we are planning to submit upstream - we deleted those because they are not part of the Oauth spec

**Comment by mutemule** [comment]
[OAuth 1.0](http://tools.ietf.org/html/rfc5849) definitely specifies only HMAC-SHA1 and RSA-SHA1, but I don't see anything along those lines in [OAuth 2.0](http://tools.ietf.org/html/rfc6749).  In some other third-party libraries, I see explicit support for HMAC-SHA256, and support for HMACs of arbitrary algorithms.

Where does OAuth 2.0 specify that only SHA1 is supported?

**Comment by kevinhughes27** [comment]
this gem is only for Oauth 1.0

**Comment by mutemule** [comment]
Well then: :+1:  As @fw42 said, this is much better.

**Comment by kevinhughes27** [comment]
@fw42 is that a :ship: from you then?

**Comment by fw42** [comment]
yep :+1:

File: .travis.yml [hunk 2]
```diff
@@ -1,9 +1,7 @@
 language: ruby
 script: "bundle exec rake test"
-sudo: false
 
 rvm:
-- "2.2"
```
Comment (shawnfrench) [inline .travis.yml]: Curious: why?

**Comment by shawnfrench** [comment]
LGTM - one question, plus not sure if we need to add byebug for upstream, but...not sure if it matters.

File: .travis.yml (hunk 2 above)
Comment (kevinhughes27) [inline .travis.yml]: travis was exploding on bundling with this ruby version so I opted to just removed it for now

**Comment by kevinhughes27** [comment]
we'll see what they say

**Pull request merged**